[Tanzhiheng26] iP

Tanzhiheng26 opened the pull request:
### Ezra

> “Your mind is for having ideas, not holding them.” – David Allen

Let Ezra liberate your mind from the burden of remembering every task. 

Why Ezra? 🤔
- Text-based
- Easy to learn
- **FREE!** 

Installation guide:
1. Download the jar file [here](https://github.com/Tanzhiheng26/ip/releases/tag/A-Jar)
2. Copy the jar file into an empty folder.
3. Open a command window in that folder.
4. Run the command java -jar ezra.jar (i.e., run the command in the same folder as the jar file).

Features:
- [ ] Managing todos
- [ ] Managing deadlines
- [ ] Managing events

Here is what the `main` method looks like:
```java
public static void main(String[] args) {
    new Ezra("data/ezra.txt").run();
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
docs/README.md: Tweak document template

Let's tweak the docs/README.md (which is used as the user guide)
to fit Duke better. Specifically,

1. mention product name in the title
2. mention adding a product screenshot and a product intro
3. tweak the flow to describe feature-by-feature

Tanzhiheng26 committed:
Rename chatbot to Ezra, print greeting and exit messages

Tanzhiheng26 committed:
Echo user input until user enters "bye", add indentation to chatbot messages

Tanzhiheng26 committed:
Store user input in tasks ArrayList and display previous inputs when user enters "list"

Tanzhiheng26 committed:
Add Task class, allow tasks to be marked as done or not done

Tanzhiheng26 committed:
Change static fields and methods to non-static.

This allows each instance of the chatbot to have its own task list.

Tanzhiheng26 committed:
Add ToDo class and handle todo command

Tanzhiheng26 committed:
Add Deadline class and handle deadline command

Tanzhiheng26 committed:
Add Event class and handle event command

Tanzhiheng26 committed:
Test UI

Tanzhiheng26 committed:
Handle commands with incorrect format using custom WrongFormatException class

Tanzhiheng26 committed:
Add delete command to remove tasks from the list

Tanzhiheng26 committed:
Add test for delete command

Tanzhiheng26 committed:
Write tasks to ./data/ezra.txt

Tanzhiheng26 committed:
Load tasks from ./data/ezra.txt

Tanzhiheng26 committed:
Create data folder if it does not exist

Tanzhiheng26 committed:
Merge branch 'branch-Level-7'

Tanzhiheng26 committed:
Store date as LocalDate in Deadline

Tanzhiheng26 committed:
Store date as LocalDate in Event

Tanzhiheng26 committed:
Store date and time as LocalDateTime in Deadline and Event

Tanzhiheng26 committed:
Merge branch 'branch-Level-8'

Tanzhiheng26 committed:
Create Ui, Storage, Parser, TaskList class and refactor Duke.java

Tanzhiheng26 committed:
Put all classes in src.main.java.ezra package

Tanzhiheng26 committed:
Put all classes in ezra package

Tanzhiheng26 committed:
Merge branch 'add-gradle-support' of https://github.com/Tanzhiheng26/ip

Tanzhiheng26 committed:
Add JUnit tests

Tanzhiheng26 committed:
Add equals method to Event, Deadline and ToDo

Tanzhiheng26 committed:
Add test case to TestParseDeadline

Tanzhiheng26 committed:
Update build.gradle to create JAR files

Tanzhiheng26 committed:
Update filepath of Ezra in runtest.bat

Tanzhiheng26 committed:
Add javadoc comments to Task and its subclasses

Tanzhiheng26 committed:
Add javadoc comments all classes

Tanzhiheng26 committed:
Follow Coding Standard

Tanzhiheng26 committed:
Add find command

Tanzhiheng26 committed:
Merge branch 'branch-A-JavaDoc'

Tanzhiheng26 committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/ezra/Storage.java
#	src/main/java/ezra/TaskList.java

Tanzhiheng26 committed:
Merge branch 'branch-Level-9'

Tanzhiheng26 committed:
Add javadoc comments for find and parseFind

Tanzhiheng26 committed:
Use '/' instead of File.separator for filepath

Tanzhiheng26 committed:
Change Task to abstract class

Tanzhiheng26 committed:
Rename toString2 to toStorageString

Tanzhiheng26 committed:
Update expected result in testcase

Tanzhiheng26 committed:
Change greet method in Ui to static

Tanzhiheng26 committed:
Configure Checkstyle and correct coding style violations

Tanzhiheng26 committed:
Merge branch 'branch-A-Checkstyle'

Tanzhiheng26 committed:
Add GUI from JavaFx tutorial

Tanzhiheng26 committed:
Integrate command line messages into GUI

Tanzhiheng26 committed:
Clip ImageView into a circle and add padding around Label

Tanzhiheng26 committed:
Change Label node to Text node within a TextFlow

This allows the chatbot to show all tasks no matter how many there are.

Tanzhiheng26 committed:
Disable maximize button

Tanzhiheng26 committed:
Remove unused imports, fields and methods in Ezra

Tanzhiheng26 committed:
Add javadoc comments to classes in gui package

Tanzhiheng26 committed:
Update javadoc comments

Tanzhiheng26 committed:
Add javadoc comment

Tanzhiheng26 committed:
Merge branch 'branch-Level-10'

Tanzhiheng26 committed:
Remove redundant variable

Tanzhiheng26 committed:
Add ezra.txt to .gitignore

Tanzhiheng26 committed:
Allow deletion of multiple tasks in 1 command

Tanzhiheng26 committed:
Allow multiple tasks to be marked/unmarked in 1 command

Tanzhiheng26 committed:
Merge branch 'branch-A-Varargs'

narwhalsilent commented on src/main/java/ezra/Event.java:
```diff
@@ -0,0 +1,83 @@
+                this.description,
+                startString,
+                endString
+        );
```
I like your formatting of the String.format method here. It is easy to read, and you comply with the 8 space continuation indent standard that many others may miss.

narwhalsilent commented on src/main/java/ezra/Ezra.java:
```diff
@@ -0,0 +1,61 @@
+    /**
+     * You should have your own function to generate a response to user input.
+     * Replace this stub with your completed method.
+     */
```
I am a little confused what this method does, and I am not completely understanding what the Javadoc is saying either. Is this something that the user should change? What if they download your release as a .jar file? If this method has been completed, it might be good to revise the Javadoc to give a more informative description of the method, including `@param` and `@return`.

Tanzhiheng26 replied:
Thanks for pointing that out, I will update the javadoc

narwhalsilent commented on src/main/java/ezra/Parser.java:
```diff
@@ -0,0 +1,170 @@
+                return tasks.find(Parser.parseFind(input));
+            } else {
+                return "Invalid command";
+            }
```
The forgoing if-else chain is formatted nicely and easy to read.

narwhalsilent commented on src/main/java/ezra/Task.java:
```diff
@@ -0,0 +1,36 @@
+     *
+     * @return A formatted string suitable for storage, including the completion status and description.
+     */
+    public abstract String toStorageString();
```
Since this is an abstract method, maybe it will be easier to keep track if you provide `@Override` tags in the child classes?

Tanzhiheng26 replied:
Good idea

narwhalsilent commented on src/main/java/ezra/TaskList.java:
```diff
@@ -0,0 +1,247 @@
+                this.tasks.remove(taskIndex);
+                hasDeletedTasks = true;
+
+                // Update tasks in storage
```
Your inline annotations are very helpful and always indented correctly.

narwhalsilent commented on src/main/java/ezra/WrongFormatException.java:
```diff
@@ -0,0 +1,16 @@
+/**
+ * Exception thrown when there is an incorrect format in user input commands.
+ */
+public class WrongFormatException extends Exception {
```
I like that you put this exception as checked, this way you will always be reminded to handle this exception.

narwhalsilent reviewed (commented):
I like that your code always follow the coding standards and it is very easy to read and understand!

Tanzhiheng26 committed:
Display greeting message

Tanzhiheng26 committed:
Add @Override to toStorageString() method

Tanzhiheng26 committed:
Update javadoc comments

Tanzhiheng26 committed:
Reorder parameters to be consistent

Tanzhiheng26 committed:
Assert that taskIndex >= -1

parseMark, parseUnmark and parseDelete ensures that only digits are
allowed after the command so the taskNumber is at least 0. taskIndex is
1 less than taskNumber so taskIndex is at least -1.

Tanzhiheng26 committed:
Delete unused methods

The run and main methods are no longer needed after adding the GUI.

Tanzhiheng26 committed:
Improve code quality in Parser

Make the happy path prominent, store complicated regex expressions in
boolean variables, extract methods

Tanzhiheng26 committed:
Update tests for parseDelete

Add test cases and change expected result from an integer to a String
array containing task numbers.

Tanzhiheng26 committed:
Improve code quality in TaskList

Use named constants, use default branch, extract methods, make the happy
path prominent.

Tanzhiheng26 committed:
Merge pull request #1 from Tanzhiheng26/branch-A-Assertions

Assert that taskIndex >= -1

Tanzhiheng26 committed:
Merge branch 'master' into branch-A-CodeQuality

Tanzhiheng26 committed:
Merge pull request #2 from Tanzhiheng26/branch-A-CodeQuality

Merge branch-A-CodeQuality to master

Tanzhiheng26 committed:
Use streams

Use streams to convert string array of taskNumbers to int array of
taskIndices.

Tanzhiheng26 committed:
Merge pull request #3 from Tanzhiheng26/branch-A-Streams

Merge branch-A-Streams to master

Tanzhiheng26 committed:
Update javadoc comments

Tanzhiheng26 committed:
Create main.yml

Tanzhiheng26 committed:
Update main.yml

Perform JUnit tests

Tanzhiheng26 committed:
Add copy method to create copies of Task objects

Tanzhiheng26 committed:
Add undo feature

Tanzhiheng26 committed:
Merge branch 'branch-C-Undo'

Tanzhiheng26 committed:
Update javadoc comments

Tanzhiheng26 committed:
Remove ./gradlew test

./gradlew check already runs all tests

Tanzhiheng26 committed:
Update profile pictures and add background color to GUI

Tanzhiheng26 committed:
Make profile pictures slightly smaller

This gives more space for the text

Tanzhiheng26 committed:
Display red text when user types wrong command

Tanzhiheng26 committed:
Add test cases for all methods in Parser

Tanzhiheng26 committed:
Add unit tests for ToDo

Tanzhiheng26 committed:
Add unit tests for equals method in Deadline

Tanzhiheng26 committed:
Add unit tests for Event

Tanzhiheng26 committed:
Merge branch 'master' of https://github.com/Tanzhiheng26/ip

Tanzhiheng26 committed:
Correct checkstyle errors

Tanzhiheng26 committed:
Remove unnecessary new line

Tanzhiheng26 committed:
Adjust margin and padding

Tanzhiheng26 committed:
Merge branch 'master' into branch-A-BetterGui

Tanzhiheng26 committed:
Merge branch 'branch-A-BetterGui'

Tanzhiheng26 committed:
Add title to application window

Tanzhiheng26 committed:
Correct checkstyle errors

Tanzhiheng26 committed:
Update README.md

Add list of features

Tanzhiheng26 committed:
Update docs/README.md

Add list of features

Tanzhiheng26 committed:
Update docs/README.md

Add Quick Start section

Tanzhiheng26 committed:
Update docs/README.md

Add screenshot and content page

Tanzhiheng26 committed:
Update docs/README.md

Reduce size of screenshot

Tanzhiheng26 committed:
Update docs/README.md

Fix screenshot not rendering in GitHub Pages

Tanzhiheng26 committed:
Update docs/README.md

Use original screenshot which has better quality

Tanzhiheng26 committed:
Add Ui.png and update README.md

Tanzhiheng26 committed:
Update javadoc comments

Tanzhiheng26 committed:
Update message format

Tanzhiheng26 committed:
Update README.md

Fix typo
